File: ruby/122-knapsack/knapsack.rb
class Knapsack
  def initialize(max_weight)
    @max_weight = max_weight
  end

  def max_value(items)
    return 0 if items.empty? || @max_weight <= 0

    n = items.length
    # Create DP table: dp[i][w] = max value using first i items with weight limit w
    dp = Array.new(n + 1) { Array.new(@max_weight + 1, 0) }

    # Fill the DP table
    (1..n).each do |i|
      item = items[i - 1]
      (0..@max_weight).each do |w|
        # Don't take the item
        dp[i][w] = dp[i - 1][w]

        # Take the item if it fits
        if item.weight <= w
          value_with_item = dp[i - 1][w - item.weight] + item.value
          dp[i][w] = [dp[i][w], value_with_item].max
        end
      end
    end

    dp[n][@max_weight]
  end
end

File: go/125-knapsack/knapsack.go
package knapsack

type Item struct {
	Weight, Value int
}

// Knapsack takes in a maximum carrying capacity and a collection of items
// and returns the maximum value that can be carried by the knapsack
// given that the knapsack can only carry a maximum weight given by maximumWeight
func Knapsack(maximumWeight int, items []Item) int {
	n := len(items)
	if n == 0 || maximumWeight == 0 {
		return 0
	}

	// Create a 2D DP table where dp[i][w] represents the maximum value
	// that can be obtained with the first i items and weight limit w
	dp := make([][]int, n+1)
	for i := range dp {
		dp[i] = make([]int, maximumWeight+1)
	}

	// Fill the DP table
	for i := 1; i <= n; i++ {
		item := items[i-1]
		for w := 0; w <= maximumWeight; w++ {
			// Option 1: Don't take the current item
			dp[i][w] = dp[i-1][w]

			// Option 2: Take the current item (if it fits)
			if item.Weight <= w {
				valueWithItem := dp[i-1][w-item.Weight] + item.Value
				if valueWithItem > dp[i][w] {
					dp[i][w] = valueWithItem
				}
			}
		}
	}

	return dp[n][maximumWeight]
}
